src/probability/martingale/optional_sampling.lean

```diff
@@ -0,0 +1,266 @@
+end
+
+/-- **Optional Sampling theorem**. If `τ` is a bounded stopping time and `σ` is another stopping
+time, then the stopped value of a martingale `f` with respect to `min τ σ` is almost everywhere
+equal to `μ[stopped_value f τ | hσ.measurable_space]`. -/
+lemma stopped_value_min_ae_eq_condexp [sigma_finite_filtration μ ℱ]
+  (h : martingale f ℱ μ) (h_prog : prog_measurable ℱ f)
+  (hτ : is_stopping_time ℱ τ) (hσ : is_stopping_time ℱ σ) {n : ι} (hτ_le : ∀ x, τ x ≤ n)
+  [h_sf_min : sigma_finite (μ.trim (hτ.min hσ).measurable_space_le)] :
+  stopped_value f (λ x, min (σ x) (τ x)) =ᵐ[μ] μ[stopped_value f τ | hσ.measurable_space] :=
+begin
+  have h_min_comm : (hτ.min hσ).measurable_space = (hσ.min hτ).measurable_space,
+    by rw [is_stopping_time.measurable_space_min, is_stopping_time.measurable_space_min, inf_comm],
+  haveI : sigma_finite (μ.trim (hσ.min hτ).measurable_space_le),
+  { convert h_sf_min; { ext1 x, rw min_comm, }, },
+  haveI : sigma_finite (μ.trim hτ.measurable_space_le),
+  { have h_le : (hτ.min hσ).measurable_space ≤ hτ.measurable_space,
+    { rw is_stopping_time.measurable_space_min,
+      exact inf_le_left, },
+    exact sigma_finite_trim_mono _ h_le, },
+  haveI : sigma_finite (μ.trim hσ.measurable_space_le),
+  { have h_le : (hτ.min hσ).measurable_space ≤ hσ.measurable_space,
+    { rw is_stopping_time.measurable_space_min,
+      exact inf_le_right, },
+    exact sigma_finite_trim_mono _ h_le, },
+  refine (h.stopped_value_ae_eq_condexp_of_le hτ (hσ.min hτ) (λ x, min_le_right _ _) hτ_le).trans _,
+  refine ae_of_ae_restrict_of_ae_restrict_compl {x | σ x ≤ τ x} _ _,
+  { exact condexp_min_stopping_time_ae_eq_restrict_le hσ hτ, },
+  { suffices : μ[stopped_value f τ|(hσ.min hτ).measurable_space]
+      =ᵐ[μ.restrict {x | τ x ≤ σ x}] μ[stopped_value f τ|hσ.measurable_space],
+    { rw ae_restrict_iff' (hσ.measurable_space_le _ (hσ.measurable_set_le_stopping_time hτ).compl),
+      rw [filter.eventually_eq, ae_restrict_iff'] at this,
+      swap, { exact hτ.measurable_space_le _ (hτ.measurable_set_le_stopping_time hσ), },
+      filter_upwards [this] with x hx hx_mem,
+      simp only [set.mem_compl_iff, set.mem_set_of_eq, not_le] at hx_mem,
+      exact hx hx_mem.le, },
+    refine filter.eventually_eq.trans _
+      ((condexp_min_stopping_time_ae_eq_restrict_le hτ hσ).trans _),
+    { exact stopped_value f τ, },
+    { rw h_min_comm, },
+    { have h1 : μ[stopped_value f τ|hτ.measurable_space] = stopped_value f τ,
+      { refine condexp_of_strongly_measurable hτ.measurable_space_le _ _,
+        { refine measurable.strongly_measurable _,
+          exact measurable_stopped_value h_prog hτ, },
+        { exact integrable_stopped_value ι hτ h.integrable hτ_le, }, },
+      rw h1,
+      exact (condexp_stopped_value_stopping_time_ae_eq_restrict_le h h_prog hτ hσ hτ_le).symm, }, },
+end
+
+/-- **Optional Sampling theorem** for martingales indexed by `ℕ`. If `τ` is a bounded stopping time
```

Comment:
If you remove `h_prog` from the previous theorem, then this one becomes a special case of the previous one and should probably be removed.